src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -96,21 +116,45 @@ public virtual bool StartSession(ITestSessionEventsHandler eventsHandler)
 
         // Wait for proxy creation to be over.
         Task.WaitAll(taskList);
+        stopwatch.Stop();
+
+        // Collecting session metrics.
+        requestData?.MetricsCollection.Add(
+            TelemetryDataConstants.TestSessionId,
+            _testSessionInfo.Id);
+        requestData?.MetricsCollection.Add(
+            TelemetryDataConstants.TestSessionSpawnedTesthostCount,
+            _proxyContainerList.Count);
+        requestData?.MetricsCollection.Add(
+            TelemetryDataConstants.TestSessionTesthostSpawnTimeInSec,
+            stopwatch.Elapsed.TotalSeconds);
 
         // Dispose of all proxies if even one of them failed during setup.
         if (_proxySetupFailed)
         {
+            requestData?.MetricsCollection.Add(
+                TelemetryDataConstants.TestSessionState,
+                TestSessionState.ProxySetupFailure.ToString());
             DisposeProxies();
             return false;
         }
 
         // Make the session available.
         if (!TestSessionPool.Instance.AddSession(_testSessionInfo, this))
         {
+            requestData?.MetricsCollection.Add(
+                TelemetryDataConstants.TestSessionState,
+                TestSessionState.SessionExistsFailure.ToString());
             DisposeProxies();
             return false;
         }
 
+        requestData?.MetricsCollection.Add(
+            TelemetryDataConstants.TestSessionState,
+            TestSessionState.Success.ToString());
+
+        _testSessionStopwatch.Start();
```

Comment:
why not include the time of "TestSessionTesthostSpawnTimeInSec" inside the session time?

Reply:
`TestSessionTesthostSpawnTimeInSec` is the actual time we spent on spawning the initial set of testhosts and we'll know this piece of info immediately after we finish the session spinoff. This is a time during which we "boot up" the session so it shouldn't be counted as part of the total session time.

The total session time (i.e. time the session was up and running) is a separate measurement that we'll have info about once someone calls `StopTestSession` on the wrapper and as such ends the session.